securedrop_client/storage.py

```diff
@@ -116,13 +118,48 @@ def update_local_storage(session: Session,
     # The following update_* functions may change the database state.
     # Because of that, each get_local_* function needs to be called just before
     # its respective update_* function.
-    update_sources(remote_sources, get_local_sources(session), session, data_dir)
+    update_sources(gpg, remote_sources, get_local_sources(session), session, data_dir)
     update_files(remote_files, get_local_files(session), session, data_dir)
     update_messages(remote_messages, get_local_messages(session), session, data_dir)
     update_replies(remote_replies, get_local_replies(session), session, data_dir)
 
 
-def update_sources(remote_sources: List[SDKSource],
+def update_source_key(
+        gpg: GpgHelper, session: Session, local_source: Source, remote_source: SDKSource
+) -> None:
+    """
+    Updates a source's GPG key.
+    """
+    if not remote_source.key.get("fingerprint"):
+        logger.error("New source data lacks key fingerprint")
+        return
+
+    if not remote_source.key.get("public"):
+        logger.error("New source data lacks public key")
+        return
+
+    if (
+        local_source.fingerprint == remote_source.key['fingerprint'] and
+        local_source.public_key == remote_source.key['public']
+    ):
+        logger.debug("Source key data is unchanged")
+        return
+
+    try:
+        # commit so the new source is visible to import_key, which uses a new session
+        session.commit()
```

Comment:
it doesn't looks like anything was changed to commit, or am i missing something? i think this works fine to commit in `update_sources`, which makes the actual changes to `local_source` before calling `update_source_key`?

Reply:
This is a good catch; I pulled this out of `update_sources` and should have left the commit where the changes are being made. It'll have to happen in two places instead of one, but they're the proper places and this will simplify the signature of `update_source_key`. I'll get to this tomorrow, but wanted to comment to make sure this didn't get merged before that's fixed.

Reply:
Done.